### ISSUE TITLE
Allow transformation of messages via Google Translate

### ISSUE THREAD

**Issue by patcon** [42wim/matterbridge#499]
**Describe the solution you'd like**
I am interested in whether matterbridge could be used to help unfragment international communities for which language barriers make collaboration difficult. It would be wonderful if matterbridge could be used to create translatable gateways between tools, or even between different channels within the same tool.

So for example, the `g0v-tw` Slack could bridge it's `#general` channel with `general-en`, and messages could be translated into each "ends" target language.

I can imagine a generalized feature request being for "transformation of messages within gateways in a directional fashion". But the specific request would be to discuss allowing translation via [Google Translate's API](https://cloud.google.com/translate/docs/reference/translate).

I could imagine this involving a pluggable system that also powered the reformatting of markup in between platforms.

**Describe alternatives you've considered**
Forking the project.

**Additional context**
None.

Thanks for your consideration!

**Comment by 42wim** [comment]
Well, it's an interesting idea. But right now, I don't have time to work on this myself.
If you know/have someone who can work on this, feel free to open a PR and I can give some pointers if necessary.

**Comment by patcon** [comment]
Thanks for vetting and encouragement :) `modifyMessage` already exists, but I'll just choose any old name for now and happy to change later

**Comment by patcon** [comment]
> Calling a new function in modifyMessage would be the best.

Ok, after digging around a bit, I'm confused by this suggestion, and was hoping you could help me understand :)

https://github.com/42wim/matterbridge/blob/296428d53e4febb5a82082d3c61628fbd396fd13/gateway/router.go#L97-L108

I appears that `modifyMessage` in `handleRecieve` is what is called based on gateway-level config that applies to every message that comes out of the gateway, and it makes changes that apply to all destination channels. It seems to be `handleMessage` that is called for each of the potential many generated messages going into other channels.

Is this correct? If so, then the latter is where the new functionality must be added, because each message must be translated differently based on the `Locale` settings of the channel it's being dropped into.

Any clarification of my understanding is appreciated! Thanks @42wim!

**Comment by 42wim** [comment]
Yes, your suggestion is correct. (sorry for sending you in the wrong direction)

**Comment by patcon** [comment]
![screen shot 2018-10-08 at 4 53 01 pm](https://user-images.githubusercontent.com/305339/46599747-b4efc200-cb1a-11e8-89b1-5677ad8dcbfb.png)

Yay! Bare working PoC!

**Comment by patcon** [comment]
This still needs some work, but I just wanted to say that this is WORKING SO WELL! I can already feel it slightly changing how our community is able to communicate, and it's pretty neat!

![screen shot 2018-10-11 at 6 17 04 pm](https://user-images.githubusercontent.com/305339/46797528-f7ff9e80-cd81-11e8-97e2-6babe860d60b.png)

Thank you so so so SO much for this tool @42wim :)))

**Comment by patcon** [comment]
Nevermind. Figured it out! Hadn't seen the `origmsg` var 🤦‍♂️
